File: daily-exercise/special_topic/array/remove-element.go
// 快慢指针
func removeElementV1(nums []int, val int) int {
    slow := 0
    for fast:=0;fast<len(nums);fast++{
        if nums[fast] != val {
            nums[slow] = nums[fast]
            slow ++
        }
    }
    return slow
}

// 前后双指针
func removeElementV2(nums []int, val int) int {
    left, right := 0, len(nums)
    for left < right {
        if nums[left] == val {
            nums[left] = nums[right-1]
            right--
        } else {
            left++
        }
    }
    return left
}

File: daily-exercise/special_topic/array/remove-element.py
class Solution:
    # 头尾双指针
    def removeElement(self, nums, val):
        left, right = 0, len(nums) - 1
        while left <= right:
            if nums[left] == val:
                nums[left] = nums[right]
                right -= 1
            else:
                left += 1
        print(nums, left, right)
        return left


# 快慢指针
def removeElement(nums, val):
    slow = 0
    for f in range(len(nums)):
        if nums[f] != val:
            nums[slow] = nums[f]
            slow += 1


test = Solution()
index = test.removeElement([0, 1, 2, 2, 3, 0, 4, 2], 2)
